### PR TITLE
Added javadoc to AbstractPlayableObjectDeserializer.java file

### PR DESCRIPTION
**Pull request by suwindev19**
Hi @jzheng2017 ,

I have added javadoc to AbstractPlayableObjectDeserializer class. Since this is my first time writing javadoc, I have got a few questions and they are listed below.  I will update this as per your feedback. Thanks very much.

1. spotify.factories package : Is this package for implementing java factory design pattern?  
2. @link , from what I read, I should only use this tag to point to another classes or methods of the project (not the official documentation). I have however added links to JSONObject documentation just in case. Should I keep or delete them?
3. I have difficulty understanding this line of code. AbstractPlayableObject childClass = null; For this code, childClass is the instance of AbstractPlayableObject which is an abstract class. I thought abstract class cannot be instantiated. 
4. I want to add @since to add version you have added for each class. When I check pom.xml, I can see API version number, 1.5.6 which I think is the latest version. To find out which version you add each class in factories package, where should I look at.
5. I would like to clarify the followings. When I commenced documentation, I read articles online so that I have a better understanding of what each class does. My understandings are as follows. Are they correct?
     AbstractPlayableObject class is a custom deserializer.

### CODE REVIEW

**Comment by jzheng2017** [comment]
Hi @suwindev19 , it's nice to see that you've picked up the task so fast :). 

To answer your questions:

> spotify.factories package : Is this package for implementing java factory design pattern?

This is indeed where the factory classes reside.

> @link , from what I read, I should only use this tag to point to another classes or methods of the project (not the official documentation). I have however added links to JSONObject documentation just in case. Should I keep or delete them?

You're right, I made a mistake. I actually meant using `<a>` tag for referering to the official documentation instead of `{@link}`

> I have difficulty understanding this line of code. AbstractPlayableObject childClass = null; For this code, childClass is the instance of AbstractPlayableObject which is an abstract class. I thought abstract class cannot be instantiated.

You're right that abstract classes can not be instantiated. But the code is not instantiating a class of `AbstractPlayableObject` but a subclass of it.

> I want to add @SInCE to add version you have added for each class. When I check pom.xml, I can see API version number, 1.5.6 which I think is the latest version. To find out which version you add each class in factories package, where should I look at.

You can look at the releases [here](https://github.com/jzheng2017/spotify-web-api-wrapper/releases) to see when the factory was added, they were most likely added in v1.0.0, but you'll have to confirm that.

> I would like to clarify the followings. When I commenced documentation, I read articles online so that I have a better understanding of what each class does. My understandings are as follows. Are they correct?
AbstractPlayableObject class is a custom deserializer.

No, `AbstractPlayableObject` is just an abstract model object, but `AbstractPlayableObjectDeseralizer` is indeed a custom json deserializer.

Also I just looked at your pull request, there are some things that are not correct yet. I'll put the global ones in this comment, and for specific one's I'll comment on the source file itself.

Global feedback:
- You have committed 27 files in total, 26 of which are files in .idea map. I have excluded them using .gitignore but they are committed anyway, they shouldn't be committed as they are IDE specific settings. This needs to be removed.
- You're working in the wrong branch, you worked in the `main` branch instead of `feature/javadoc` branch. The javadoc must be done in `feature/javadoc`
- The commit message does not conform to the [git guidelines](https://github.com/jzheng2017/spotify-web-api-wrapper/blob/main/CONTRIBUTING.md#commit-messages) as defined in the contribution guidelines. This needs to be changed.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java [hunk 1]
```diff
@@ -7,13 +7,46 @@
 
 import java.lang.reflect.Type;
 
+/**
+ * This is a AbstractPlayableObjectDeserializer class to implement JsonDeserializer.
```
Comment (jzheng2017) [inline src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:11]: This actually just repeats this part `class AbstractPlayableObjectDeserializer implements JsonDeserializer<AbstractPlayableObject>` but doesn't really describe what the class is for. Javadoc should describe what the class does and what it is used for.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java [hunk 2]
```diff
@@ -7,13 +7,46 @@
...
+ *
+ * @see spotify.models.generic.AbstractPlayableObject
+ * @author Jiankai Zheng (jk.zheng@hotmail.com)
+ * @since // check with Jiankai
+ */
+
 class AbstractPlayableObjectDeserializer implements JsonDeserializer<AbstractPlayableObject> {
+
+    /**
+     * Custom deserialize method to deserialize json, type and jsonDeserializationContext
```
Comment (jzheng2017) [inline src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:21]: This doesn't really also make it any clearer, it should mention something about that it deserializes based on the type returned in the json

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java [hunk 3]
```diff
@@ -7,13 +7,46 @@
...
+     * @param json represents a json element
+     * @param type represents various types
+     * @param jsonDeserializationContext
+     * @return childClass
```
Comment (jzheng2017) [inline src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:25]: Not descriptive enough, child class of what?

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java [hunk 4]
```diff
@@ -7,13 +7,46 @@
...
+ * @since // check with Jiankai
+ */
+
 class AbstractPlayableObjectDeserializer implements JsonDeserializer<AbstractPlayableObject> {
+
+    /**
+     * Custom deserialize method to deserialize json, type and jsonDeserializationContext
+     * @param json represents a json element
+     * @param type represents various types
+     * @param jsonDeserializationContext
```
Comment (jzheng2017) [inline src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:24]: There is no description

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java [hunk 5]
```diff
@@ -7,13 +7,46 @@
...
+ * @author Jiankai Zheng (jk.zheng@hotmail.com)
+ * @since // check with Jiankai
+ */
+
 class AbstractPlayableObjectDeserializer implements JsonDeserializer<AbstractPlayableObject> {
+
+    /**
+     * Custom deserialize method to deserialize json, type and jsonDeserializationContext
+     * @param json represents a json element
+     * @param type represents various types
```
Comment (jzheng2017) [inline src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:23]: What types?

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java [hunk 6]
```diff
@@ -7,13 +7,46 @@
...
+ * @see spotify.models.generic.AbstractPlayableObject
+ * @author Jiankai Zheng (jk.zheng@hotmail.com)
+ * @since // check with Jiankai
+ */
+
 class AbstractPlayableObjectDeserializer implements JsonDeserializer<AbstractPlayableObject> {
+
+    /**
+     * Custom deserialize method to deserialize json, type and jsonDeserializationContext
+     * @param json represents a json element
```
Comment (jzheng2017) [inline src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:22]: What json element does get passed in?

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java [hunk 7]
```diff
@@ -7,13 +7,46 @@
...
+     * @param type represents various types
+     * @param jsonDeserializationContext
+     * @return childClass
+     * @throws JsonParseException (RuntimeException) if the incoming Json is bad/malicious/ serious issues that occurs during parsing of a Json string.
+     */
+
     @Override
     public AbstractPlayableObject deserialize(JsonElement json, Type type, JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
+        /**
```
Comment (jzheng2017) [inline src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:31]: Writing javadoc for every line of code is way too extreme, this is not needed. Same goes for the javadoc under this line of code

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 1 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:11]: Hi @jzheng2017 ,

Thanks for your  feedback. Would this description be better?

This is a AbstractPlayableObjectDeserializer class which provides the deserialize method to deserialize json, type and jsonDeserializationContext. Thanks.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 2 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:21]: Would the following be better? Thanks.

Deserialize json, type and jsonDeserializationContext and return childClass which either contains track or episode

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 3 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:25]: Would this be better?

childClass that deserialize JSON string either "track" or "episode" into Java objects

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 4 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:24]: Would this be better?

- @param json represents the json element to deserialize
- @param type the type to convert the data into
- @param jsonDeserializationContext the context in which are the elements will be deserialized

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 5 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:23]: I have included in above.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 6 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:22]: types are track and episode. I have included in one of the above comments.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 7 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:31]: Thanks for your thorough feedback! I really appreciate it.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 5 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:23]: Thanks for your thorough feedback!

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 1 above)
Comment (jzheng2017) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:11]: It's better than what it was before, but not quite. It should explain _what_ the class does without going too much into the implementation. 

So a rough description would be like, `This class is a custom {@link JsonDeserializer} which deserializes json and returns a {@link AbstractPlayableObject} subclass based on the type property in the json `

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 2 above)
Comment (jzheng2017) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:21]: No, this is just the same but the order of words are just rearranged. The description should tell in simple words what the function is doing.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 3 above)
Comment (jzheng2017) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:25]: No it should tell that it's returning a object which is a subclass of AbstractPlayableObject

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 4 above)
Comment (jzheng2017) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:24]: - no, the first param should just say 'the json to be deserialized'
- second param is fine
- third param, context for deserialization that is passed to a custom deserializer during invocation of its JsonDeserializer.deserialize method

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 1 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:11]: Thanks. I will update it as per your feedback.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 4 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:24]: Thanks. I will update them.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 3 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:25]: Thanks. I will update the documentation.

File: src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java (hunk 2 above)
Comment (suwindev19) [reply src/main/java/spotify/factories/AbstractPlayableObjectDeserializer.java:21]: Thanks. I will try again and will include the updated one in the next PR.

**Comment by jzheng2017** [comment]
Since you have a new pul request, this one will be closed.

**Pull request closed without merging**